lib/Sema/CSSimplify.cpp

```diff
@@ -4340,8 +4340,34 @@ bool ConstraintSystem::repairFailures(
     // as a narrow exception to SE-0110, see `matchFunctionTypes`.
     //
     // But if `T.Element` didn't get resolved to `Void` we'd like
-    // to diagnose this as a missing argument which can't be ignored.
+    // to diagnose this as a missing argument which can't be ignored or
+    // a tuple is trying to be inferred as a tuple for destructuring but
+    // contextual argument does not match(in this case we remove the extra
+    // closure arguments).
     if (arg != getTypeVariables().end()) {
+      if (auto argToParamElt =
+              path.back().getAs<LocatorPathElt::ApplyArgToParam>()) {
```

Comment:
Can we detect that earlier during `matchFunctionTypes` for closure argument? That would make it much easier to diagnose as well because we'd have a type of parameter as right-hand side there.

Reply:
Hey @xedin :)
Yeah, I can check that for sure. But as far as I saw the reason this fix is currently not attempted on  `matchFunctionTypes` is because of some logic of imploding parameters, I guess related to SE-110 logic but not sure.
I'll take a deeper look and see if we can somehow that to detect and record this there

Reply:
Sounds great! IHMO that would be a much cleaner and better fix, doing so here seems too late to me.

Reply:
Quick question, it seems that we arrive at `matchFunctionTypes` for the parameters with arg function with 2 args (x,y) and parameter fn with 1 arg. But after imploding both function params become one. So my question is if is somehow possible to know that destructuring will fail at this point (because we don't have fixed types and it only adds a subtype constraint)? So the only place we know that we fail is here where the subtype constraint fails and end up here in `repairFailures` where this fix is being recorded. 
For what I could understand, it is possible to know this by function params count diff when not imploding, but when imploding happens we only know this fails at this point(subtype constraint fails). With that said, I can't think in a way to do this... :( 
Do you have any advice if this is possible, how we can detect this in `matchFunctionTypes`?

Reply:
For what I could see this is done so code like 
```swift
func f( _ param: ((Int, Int)) -> ()) {}
f( { x, y in  })
```
Could still be valid... is that correct?

Reply:
Wouldn't this example be matching `($T_x, $T_y) -> $Result` to `((Int, Int)) -> Void`? Closures aren't resolved until either it's possible to determine a contextual type or there is definitely no such type.

Reply:
Hum, not sure... I may just be confused by this parameter implode logic and with debugging couldn't see a fixed type on `matchFunctionTypes`. Here is the output of both functions
```swift
(lldb) p func1->dump()
(function_type escaping
  (input=function_params num_params=2
    (param internal_name=x
      (type_variable_type id=7))
    (param internal_name=y
      (type_variable_type id=8)))
  (output=type_variable_type id=9))
(lldb) p func2->dump()
(function_type throws
  (input=function_params num_params=1
    (param
      (type_alias_type decl=Swift.(file).Range extension.Element underlying='_'
        (parent=bound_generic_struct_type decl=Swift.(file).Range
          (type_variable_type id=1)))))
  (output=type_variable_type id=12))
(lldb) p getFixedTypeRecursive(func1, true)->dump()
(function_type escaping
  (input=function_params num_params=2
    (param internal_name=x
      (type_variable_type id=7))
    (param internal_name=y
      (type_variable_type id=8)))
  (output=type_variable_type id=9))
(lldb) p getFixedTypeRecursive(func2, true)->dump()
(function_type throws
  (input=function_params num_params=1
    (param
      (type_alias_type decl=Swift.(file).Range extension.Element underlying='_'
        (parent=bound_generic_struct_type decl=Swift.(file).Range
          (type_variable_type id=1)))))
  (output=type_variable_type id=12))
```
But after imploding both `func1Params` and `func2Params` become list with a single parameter... so that's the point I got stucked.
Sorry for all the questions and thanks for the review as always :)

Reply:
The `(type_variable_type id=1)` is yet to be bound because it may depend on the closure since the values are literals?

Reply:
Okay, that makes sense in this case parameter side haven't been resolved yet so it's still a `DependentMemberType` with unresolved base. Hm, I guess we have no other cause but to wait until both sides are solved so we'd have to do that in `repairFailures` for `FunctionArgument`, I need to think about this some more...

Reply:
Ah ok, no problem! And thank you for the help =]